File: java/easy/0145_binary_tree_postorder_traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    Wrapper wrapper = new Wrapper();
    public List<Integer> postorderTraversal(TreeNode root) {  
        traversal(root);
        return wrapper.list;
    }
    
    public void traversal(TreeNode root) {
        if (root == null) return;
        traversal(root.left);
        traversal(root.right);
        wrapper.list.add(root.val);
        
    }
}

class Wrapper {
    public ArrayList<Integer> list;
    public Wrapper() {
        list = new ArrayList<>();
    }
}
/**
* Runtime: 0 ms, faster than 100.00% of Java online submissions for Binary Tree Postorder Traversal.
* Memory Usage: 39.2 MB, less than 6.31% of Java online submissions for Binary Tree Postorder Traversal.
*/

File: rust/easy/0145_binary_tree_postorder_traversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn postorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        fn dfs(root: &Option<Rc<RefCell<TreeNode>>>, answer: &mut Vec<i32>) {
            match root {
                None => return,
                Some(root) => {
                    let root = root.borrow();
                    dfs(&root.left, answer);
                    dfs(&root.right, answer);
                    answer.push(root.val);
                }
            }
        }
        let mut answer = vec![];
        dfs(&root, &mut answer);
        answer
    }
}
